crates/pathfinder/src/storage.rs

```diff
@@ -162,85 +162,136 @@ mod tests {
+            let transaction = conn.transaction().unwrap();
+
+            migrate_database(&transaction).unwrap();
+            let version = schema_version(&transaction).unwrap();
+            assert_eq!(version, DB_VERSION_CURRENT);
+        }
+
+        #[test]
+        fn from_v1() {
+            let mut conn = rusqlite::Connection::open_in_memory().unwrap();
+            let transaction = conn.transaction().unwrap();
+
+            // Set up a v1 db
+            enable_foreign_keys(&transaction).unwrap();
+            transaction
+                .pragma_update(None, VERSION_KEY, DB_VERSION_1)
+                .unwrap();
+            transaction
+                .execute(
+                    r"CREATE TABLE contracts (
+                            address    BLOB PRIMARY KEY,
+                            hash       BLOB NOT NULL,
+                            bytecode   BLOB,
+                            abi        BLOB,
+                            definition BLOB
+                        )",
+                    [],
+                )
+                .unwrap();
+            transaction
+                .execute(
+                    r"CREATE TABLE ethereum_blocks (
+                            hash   BLOB PRIMARY KEY,
+                            number INTEGER NOT NULL
+                        )",
+                    [],
+                )
+                .unwrap();
+            transaction
+                .execute(
+                    r"CREATE TABLE ethereum_transactions (
+                            hash       BLOB PRIMARY KEY,
+                            idx        INTEGER NOT NULL,
+                            block_hash BLOB NOT NULL,
+        
+                            FOREIGN KEY(block_hash) REFERENCES ethereum_blocks(hash)
+                        )",
+                    [],
+                )
+                .unwrap();
+            transaction.execute(
+                    r"CREATE TABLE global_state (
+                            starknet_block_hash       BLOB PRIMARY KEY,
+                            starknet_block_number     INTEGER NOT NULL,
+                            starknet_block_timestamp  INTEGER NOT NULL,
+                            starknet_global_root      BLOB NOT NULL,
+                            ethereum_transaction_hash BLOB NOT NULL,
+                            ethereum_log_index        INTEGER NOT NULL,
+        
+                            FOREIGN KEY(ethereum_transaction_hash) REFERENCES ethereum_transactions(hash)
+                        )",
+                    [],
+                ).unwrap();
+            transaction
+                .execute(
+                    r"CREATE TABLE contract_states (
+                            state_hash BLOB PRIMARY KEY,
+                            hash       BLOB NOT NULL,
+                            root       BLOB NOT NULL
+                        )",
+                    [],
+                )
+                .unwrap();
```

Comment:
I don't think you should be doing this "migration" manually in a test. What I mean is there should somewhere be a function that migrates from `VERSION=0` to `VERSION=1`. And then again from `VERSION=1` to `VERSION=2`. And in general `VERSION=N` to `VERSION=N+1`.

This means our total migration function would always be a loop:
```rust
for from_version in current_version..target_version {
    match from_version {
        0 => migrate_from_0_to_1();
        1 => migrate_from_1_to_2();
        2 => migrate_from_2_to_3();
        // etc
    }
}
```
Unless we do it this way, I don't see how we would ever manage this process correctly across multiple versions. Thoughts?

Reply:
1. Yes, the version bump should not really be merged into main, as the schema is not stabilized and what I called "v1" is going to be an artificial step that we will never come across again.
2. As for testing - the problem here was the I did not have any raw data to test against and yet wanted to add a test case for v1 to latest migration, so I hand crafted one. Having test fixtures as db files for each version could be some alternative to manually checking the schema.
3. Migrating in a cascade is the only way to go ofc, otherwise permutations explode. I left the flow for 0->2/latest as it will remain anyway. 0->1 is trivial and would be thrown away, and only 1->2 was something more interesting to look at, especially with the real db dump that we have to try on.
4. The next step I propose is:
- remove `_V1` and everything related to it,
- extract empty to latest migration to a separate `0_to_1` fn and add the cascading loop/logic
- `migrate` steps should also indicate if vacuuming is required afterwards

Reply:
Sounds good to me 👍 I know this was just a practice run, just wanted to make sure we were on the same page regarding the future.